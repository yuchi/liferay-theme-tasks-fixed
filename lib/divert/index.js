'use strict';

let _ = require('lodash');

let lfrThemeConfig = require('../liferay_theme_config');

function divert(moduleName, version = null) {
	if (!version) {
		let config = lfrThemeConfig.getConfig();

		version = config ? config.version : '7.1';
	}

	let module = {};

	module = Object.assign(module, safeRequire(`./common/${moduleName}`));
	module = Object.assign(module, safeRequire(`./${version}/${moduleName}`));

	return module;
}

function safeRequire(moduleName) {
	let module = {};

	try {
		module = require(moduleName);
	} catch (err) {
		if (err.message !== `Cannot find module '${moduleName}'`) {
			throw err;
		}
	}

	if (!_.isPlainObject(module)) {
		throw new Error(
			'Only modules exporting plain objects can be used with divert'
		);
	}

	return module;
}

module.exports = divert;
