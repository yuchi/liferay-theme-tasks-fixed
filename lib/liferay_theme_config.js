const _ = require('lodash');
const fs = require('fs-extra');
const path = require('path');

function getConfig(all, alternatePath) {
	let packageJSON = getPackageJSON(alternatePath);

	return all ? packageJSON : packageJSON.liferayTheme;
}

function removeConfig(settings) {
	let packageJSON = getPackageJSON();

	packageJSON.liferayTheme = _.omit(packageJSON.liferayTheme, settings);

	writePackageJSON(packageJSON);
}

function removeDependencies(dependencies) {
	let packageJSON = getPackageJSON();

	deleteDependencies(packageJSON.dependencies, dependencies);
	deleteDependencies(packageJSON.devDependencies, dependencies);

	writePackageJSON(packageJSON);
}

function setConfig(config) {
	let packageJSON = getPackageJSON();

	if (packageJSON.liferayTheme) {
		if (config.baseTheme) {
			packageJSON.liferayTheme.baseTheme = undefined;
		}

		if (config.themeletDependencies) {
			packageJSON.liferayTheme.themeletDependencies = undefined;
		}
	}

	packageJSON = _.merge(packageJSON, {
		liferayTheme: config,
	});

	writePackageJSON(packageJSON);
}

function setDependencies(dependencies, devDependencies) {
	let packageJSON = getPackageJSON();

	let selector = devDependencies ? 'devDependencies' : 'dependencies';

	if (!packageJSON[selector]) {
		packageJSON[selector] = {};
	}

	_.merge(packageJSON[selector], dependencies);

	writePackageJSON(packageJSON);
}

module.exports = {
	getConfig,
	removeConfig,
	removeDependencies,
	setConfig,
	setDependencies,
};

function deleteDependencies(sourceDependencies, deletedDependencies) {
	_.forEach(sourceDependencies, function(item, index) {
		if (deletedDependencies.indexOf(index) > -1) {
			delete sourceDependencies[index];
		}
	});
}

function getPackageJSON(alternatePath) {
	alternatePath = alternatePath || process.cwd();

	let packageJSONContent = fs.readFileSync(
		path.join(alternatePath, 'package.json'),
		{
			encoding: 'utf8',
		}
	);

	return JSON.parse(packageJSONContent);
}

function writePackageJSON(json) {
	fs.writeFileSync(
		path.join(process.cwd(), 'package.json'),
		JSON.stringify(json, null, '\t')
	);
}
