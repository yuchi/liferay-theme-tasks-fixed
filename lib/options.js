const _ = require('lodash');
const minimist = require('minimist');

const lfrThemeConfig = require('./liferay_theme_config');

let options;

function getOptions(config) {
	if (!options || config) {
		config = config || {};
		config.argv = minimist(process.argv.slice(2));
		config.pathBuild = config.pathBuild || './build';
		config.pathDist = config.pathDist || './dist';
		config.pathSrc = config.pathSrc || './src';
		config.rubySass = config.rubySass || false;
		config.sassOptions = config.sassOptions || {};

		let themeConfig = lfrThemeConfig.getConfig();

		_.assign(config, themeConfig);

		options = config;
	}

	return options;
}

module.exports = getOptions;
