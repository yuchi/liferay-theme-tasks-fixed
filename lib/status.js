const _ = require('lodash');
const colors = require('ansi-colors');

function status(themeConfig) {
	let statusBuffer = [];

	let baseTheme = themeConfig.baseTheme;

	let baseThemeLabel = colors.cyan('Base theme: ');

	if (baseTheme) {
		let baseThemeName = baseTheme.name || baseTheme;
		let baseThemeVersion = baseTheme.version ? 'v' + baseTheme.version : '';

		statusBuffer.push(
			baseThemeLabel +
				colors.green(`${baseThemeName} ${baseThemeVersion}`)
		);
	} else {
		statusBuffer.push(
			baseThemeLabel + colors.red('no base theme specified')
		);
	}

	let themeletDependencies = themeConfig.themeletDependencies;

	if (themeletDependencies) {
		statusBuffer.push(colors.cyan('Themelets:'));

		_.forEach(themeletDependencies, function(item) {
			statusBuffer.push(
				' - ' + colors.green(item.name + ' v' + item.version)
			);
		});
	}

	return statusBuffer.join('\n') + '\n';
}

module.exports = status;
