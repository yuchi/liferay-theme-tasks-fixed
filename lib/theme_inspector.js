const path = require('path');

const TEMPLATE_LANG_FTL = 'ftl';
const TEMPLATE_LANG_VM = 'vm';

function getBaseThemeGlob(themePath) {
	const templateLanguage = getLiferayThemeJSON(themePath).templateLanguage;
	const glob = '**/!(package.json';

	switch (templateLanguage) {
	case TEMPLATE_LANG_FTL:
		return glob + '|*.' + TEMPLATE_LANG_VM + ')';

	case TEMPLATE_LANG_VM:
		return glob + '|*.' + TEMPLATE_LANG_FTL + ')';

	default:
		return glob + ')';
	}
}

function getLiferayThemeJSON(themePath) {
	return require(path.join(themePath, 'package.json')).liferayTheme;
}

module.exports = {getBaseThemeGlob, getLiferayThemeJSON};
