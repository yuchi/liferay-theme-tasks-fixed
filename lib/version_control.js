'use strict';

module.exports = function() {
	try {
		let colors = require('ansi-colors');
		let updateNotifier = require('update-notifier');

		let notifier = updateNotifier({
			pkg: require('../package.json'),
			updateCheckInterval: 1000,
		});

		let update = notifier.update;

		if (update) {
			let message = [
				'You are running an old version of ' + colors.cyan(update.name),
				'Run ' +
					colors.cyan('npm update', update.name) +
					' to install newest version',
				'Current version: ' +
					colors.green(update.current) +
					' Latest version: ' +
					colors.green(update.latest),
			].join('\n');

			notifier.notify({
				message: message,
			});
		}
	} catch (err) {
		process.stdout.write(err);
	}
};
