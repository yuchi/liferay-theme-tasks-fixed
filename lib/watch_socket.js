const _ = require('lodash');
const del = require('del');
const GogoShell = require('gogo-shell');
const os = require('os');
const path = require('path');

const lfrThemeConfig = require('./liferay_theme_config');
const themeUtil = require('./util');

const themeConfig = lfrThemeConfig.getConfig(true);

const REGEX_WIN = /^win/;

class WatchSocket extends GogoShell {
	constructor(config) {
		super(config);

		config = config || {};

		this.webBundleDir = config.webBundleDir || '.web_bundle_build';
		this.deploymentStrategy = config.deploymentStrategy;
		this.dockerContainerName = config.dockerContainerName;
		this.dockerThemePath = config.dockerThemePath;
	}

	deploy() {
		return this._getWebBundleData()
			.then(data => {
				return data.id ? this._uninstallBundle(data.id) : data;
			})
			.then(this._installWebBundleDir.bind(this))
			.then(data => {
				const webBundleId = this._getWebBundleIdFromResponse(data);

				return this._startBundle(webBundleId);
			})
			.then(() => {
				this.end();
			});
	}

	uninstall(warPath, distName) {
		if (this._isDocker()) {
			const delPath = themeUtil
				.dockerExec(this.dockerContainerName, 'ls ' + warPath)
				.stdout.toString();

			if (!delPath.length) {
				return;
			}

			themeUtil.dockerExec(this.dockerContainerName, 'rm -rf ' + warPath);
		} else {
			const delPath = del.sync(warPath, {
				dryRun: true,
				force: true,
			});

			if (!delPath.length) {
				return;
			}

			del.sync(warPath, {
				force: true,
			});
		}

		return this._waitForUninstall(distName);
	}

	_formatWebBundleDirCommand(themePath) {
		let buildPath = path.join(themePath, this.webBundleDir);

		buildPath = buildPath.split(path.sep).join('/');

		if (!this._isWin() || this._isDocker()) {
			buildPath = '/' + buildPath;
		}

		buildPath = buildPath.replace(/\s/g, '%20');

		const themeName = themeConfig.name;

		return (
			'install \'webbundledir:file:/' +
			buildPath +
			'?Web-ContextPath=/' +
			themeName +
			'\''
		);
	}

	_getWebBundleData() {
		const themeName = themeConfig.name;

		const grepRegex = '\'webbundle(dir|):file.*' + themeName + '\'';

		return this.sendCommand('lb -u | grep ' + grepRegex).then(data => {
			const lines = data.split('\n');

			const result = lines[1];

			return this._getWebBundleDataFromResponse(result);
		});
	}

	_getWebBundleDataFromResponse(response) {
		let data = {
			status: null,
		};

		if (!response) {
			return data;
		}

		let i = response.indexOf('webbundle:file');

		if (i == -1) {
			i = response.indexOf('webbundledir:file');
		}

		if (i > -1) {
			const fields = response.split('|');

			data = {
				id: _.trim(fields[0]),
				level: _.trim(fields[2]),
				status: _.trim(fields[1]),
				updateLocation: _.trim(fields[3]),
			};
		}

		return data;
	}

	_getWebBundleIdFromResponse(response) {
		const match = response.match(/Bundle\sID:\s*([0-9]+)/);

		return match ? match[1] : 0;
	}

	_installWebBundleDir() {
		return this.sendCommand(
			this._formatWebBundleDirCommand(
				this._isDocker() ? this.dockerThemePath : process.cwd()
			)
		);
	}

	_isWin() {
		return REGEX_WIN.test(os.platform());
	}

	_isDocker() {
		return this.deploymentStrategy === 'DockerContainer';
	}

	_startBundle(bundleId) {
		return this.sendCommand('start', bundleId);
	}

	_stopBundle(bundleId) {
		return this.sendCommand('stop', bundleId);
	}

	_uninstallBundle(bundleId) {
		return this.sendCommand('uninstall', bundleId);
	}

	_waitForUninstall(distName) {
		return this.sendCommand('lb ' + distName)
			.delay(200)
			.then(data => {
				if (data.indexOf('No matching bundles found') < 0) {
					return this._waitForUninstall(distName);
				}
			});
	}
}

module.exports = WatchSocket;
