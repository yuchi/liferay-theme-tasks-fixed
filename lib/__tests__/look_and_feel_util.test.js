const _ = require('lodash');
const fs = require('fs-extra');
const path = require('path');
const sinon = require('sinon');

const testUtil = require('../../test/util');

const baseLookAndFeelJSON = require('./fixtures/look_and_feel_util/base-look-and-feel.json');
const mixedLookAndFeelJSON = require('./fixtures/look_and_feel_util/mixed-look-and-feel.json');
const parentLookAndFeelJSON = require('./fixtures/look_and_feel_util/parent-look-and-feel.json');

const initCwd = process.cwd();
const baseThemePath = path.join(
	__dirname,
	'../../test/fixtures/themes/7.0/base-theme'
);

let lookAndFeelUtil;

beforeEach(() => {
	process.chdir(baseThemePath);

	require('../../lib/options')({
		pathSrc: './custom_src_path',
	});

	lookAndFeelUtil = require('../look_and_feel_util.js');
});

afterEach(() => {
	process.chdir(initCwd);
});

it('lookAndFeelUtil should sort json correctly and build valid xml', () => {
	const xml = testUtil.stripNewlines(
		lookAndFeelUtil.buildXML(
			mixedLookAndFeelJSON,
			lookAndFeelUtil.getLookAndFeelDoctypeByVersion('7.0')
		)
	);

	expect(xml).toMatchSnapshot();
});

it('correctJSONIdentifiers should set the theme element id and name attribute if the name argument is different - for themes that do not have a look-and-feel.xml', () => {
	lookAndFeelUtil.correctJSONIdentifiers(baseLookAndFeelJSON, 'Test Name');

	expect(_.get(baseLookAndFeelJSON, 'look-and-feel.theme.0.$.id')).toBe(
		'test-name'
	);
	expect(_.get(baseLookAndFeelJSON, 'look-and-feel.theme.0.$.name')).toBe(
		'Test Name'
	);

	lookAndFeelUtil.correctJSONIdentifiers(baseLookAndFeelJSON, 'Base Theme');

	expect(_.get(baseLookAndFeelJSON, 'look-and-feel.theme.0.$.id')).toBe(
		'base-theme'
	);
	expect(_.get(baseLookAndFeelJSON, 'look-and-feel.theme.0.$.name')).toBe(
		'Base Theme'
	);
});

it('getLookAndFeelDoctype should extract doctype from liferay-look-and-feel.xml', () => {
	const doctype = lookAndFeelUtil.getLookAndFeelDoctype(baseThemePath);

	expect(doctype).toBe(
		'<!DOCTYPE look-and-feel PUBLIC "-//Liferay//DTD Look and Feel 7.0.0//EN" "http://www.liferay.com/dtd/liferay-look-and-feel_7_0_0.dtd">'
	);
});

it('getLookAndFeelDoctypeByVersion should generate correct doctype string based on version', () => {
	const doctype = lookAndFeelUtil.getLookAndFeelDoctypeByVersion('1.0');

	expect(doctype).toBe(
		'<!DOCTYPE look-and-feel PUBLIC "-//Liferay//DTD Look and Feel 1.0.0//EN" "http://www.liferay.com/dtd/liferay-look-and-feel_1_0_0.dtd">'
	);
});

it('getLookAndFeelJSON should convert look-and-feel xml to json', done => {
	lookAndFeelUtil.getLookAndFeelJSON(baseThemePath, result => {
		expect(result).toBeTruthy();
		expect(result).toHaveProperty('look-and-feel');
		expect(result['look-and-feel']).toBeTruthy();

		done();
	});
});

it('mergeLookAndFeelJSON should merge look and feel json', done => {
	lookAndFeelUtil.mergeLookAndFeelJSON(baseThemePath, {}, result => {
		expect(result).toBeTruthy();
		expect(result).toHaveProperty('look-and-feel');

		expect(result).toMatchSnapshot();

		done();
	});
});

it('readLookAndFeelXML should return xml and access file only once', () => {
	// erase cache for sinon spy
	Object.keys(lookAndFeelUtil.xmlCache).forEach(
		key => delete lookAndFeelUtil.xmlCache[key]
	);

	const spy = sinon.spy(fs, 'readFileSync');

	let xml = lookAndFeelUtil.readLookAndFeelXML(baseThemePath);

	xml = lookAndFeelUtil.readLookAndFeelXML(baseThemePath);

	expect(/<look-and-feel>/.test(xml)).toBeTruthy();
	expect(spy.callCount).toBe(1);
});

it('extractThemeElement should extract elements based on tag name', done => {
	lookAndFeelUtil.getLookAndFeelJSON(baseThemePath, result => {
		let portletDecorators = lookAndFeelUtil.extractThemeElement(
			result,
			'portlet-decorator'
		);

		expect(portletDecorators.length).toBe(2);
		expect(portletDecorators[0].$.id).toBe('portlet-decorator-1');
		expect(portletDecorators[1].$.id).toBe('portlet-decorator-2');

		done();
	});
});

it('mergeJSON should merge look-and-feel json and output a valid look-and-feel json object', () => {
	let lookAndFeelJSON = lookAndFeelUtil.mergeJSON(
		baseLookAndFeelJSON,
		baseLookAndFeelJSON
	);

	expect(lookAndFeelJSON).toEqual(baseLookAndFeelJSON);

	lookAndFeelJSON = lookAndFeelUtil.mergeJSON(
		baseLookAndFeelJSON,
		parentLookAndFeelJSON
	);

	expect(lookAndFeelJSON).toMatchSnapshot();
});

it('mergeThemeElementById should return null if both arrays are null', () => {
	let mergedElements = lookAndFeelUtil.mergeThemeElementById(null, null);

	expect(mergedElements).toBeNull();
});

it('mergeThemeElementById should return non null array if one of the arrays is null', () => {
	let elements1 = [
		{
			$: {
				id: 'one',
			},
		},
	];
	let elements2 = [
		{
			$: {
				id: 'two',
			},
		},
	];

	let mergedElements = lookAndFeelUtil.mergeThemeElementById(elements1, null);

	expect(mergedElements).toBe(elements1);

	mergedElements = lookAndFeelUtil.mergeThemeElementById(null, elements2);

	expect(mergedElements).toBe(elements2);
});

it('mergeThemeElementById should merge arrays and exclude repeated elements according to id', () => {
	let elements1 = [
		{
			$: {
				id: 'one',
			},
		},
		{
			$: {
				id: 'two',
			},
		},
	];
	let elements2 = [
		{
			$: {
				id: 'two',
			},
		},
		{
			$: {
				id: 'three',
			},
		},
	];

	let mergedElements = lookAndFeelUtil.mergeThemeElementById(
		elements1,
		elements2
	);

	expect(mergedElements).toEqual([
		{
			$: {
				id: 'one',
			},
		},
		{
			$: {
				id: 'two',
			},
		},
		{
			$: {
				id: 'three',
			},
		},
	]);
});

it('mergeThemeElementByValue should return null if both arrays are null', () => {
	let mergedElements = lookAndFeelUtil.mergeThemeElementByValue(null, null);

	expect(mergedElements).toBeNull();
});

it('mergeThemeElementByValue should return non null array if one of the arrays is null', () => {
	let elements1 = ['one', 'two'];
	let elements2 = ['two', 'three'];

	let mergedElements = lookAndFeelUtil.mergeThemeElementByValue(
		elements1,
		null
	);

	expect(mergedElements).toEqual(elements1);

	mergedElements = lookAndFeelUtil.mergeThemeElementByValue(null, elements1);

	expect(mergedElements).toEqual(elements1);
});

it('mergeThemeElementByValue should merge arrays and exclude repeated elements', () => {
	let elements1 = ['one', 'two'];
	let elements2 = ['two', 'three'];

	let mergedElements = lookAndFeelUtil.mergeThemeElementByValue(
		elements1,
		elements2
	);

	expect(mergedElements).toEqual(['one', 'two', 'three']);
});
