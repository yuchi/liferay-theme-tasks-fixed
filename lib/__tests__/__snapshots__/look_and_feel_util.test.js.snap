// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lookAndFeelUtil should sort json correctly and build valid xml 1`] = `"<?xml version=\\"1.0\\"?><!DOCTYPE look-and-feel PUBLIC \\"-//Liferay//DTD Look and Feel 7.0.0//EN\\" \\"http://www.liferay.com/dtd/liferay-look-and-feel_7_0_0.dtd\\"><look-and-feel>	<compatibility>		<version>7.0.0+</version>	</compatibility>	<theme id=\\"linked-70\\" name=\\"linked-70\\">		<template-extension>vm</template-extension>		<settings>			<setting configurable=\\"true\\" key=\\"show-breadcrumb\\" type=\\"checkbox\\" value=\\"true\\"/>			<setting configurable=\\"false\\" key=\\"parent-setting\\" type=\\"input\\" value=\\"false\\"/>		</settings>		<roles>			<role-name>Owner</role-name>		</roles>		<color-scheme id=\\"01\\" name=\\"Democrat\\">			<css-class>democrat</css-class>		</color-scheme>		<color-scheme id=\\"02\\" name=\\"Republican\\">			<css-class>republican</css-class>		</color-scheme>		<color-scheme id=\\"03\\" name=\\"Independent\\">			<css-class>independent</css-class>		</color-scheme>		<color-scheme id=\\"04\\" name=\\"Cool Colors\\">			<css-class>cool-colors</css-class>		</color-scheme>		<layout-templates>			<standard>				<layout-template id=\\"parent-template\\">					<template-path>/layouttpl/standard/max.tpl</template-path>					<wap-template-path>/layouttpl/standard/max.wap.tpl</wap-template-path>					<thumbnail-path>/layouttpl/standard/max.png</thumbnail-path>				</layout-template>			</standard>		</layout-templates>		<portlet-decorator id=\\"linked\\" name=\\"Linked\\">			<portlet-decorator-css-class>portlet-linked</portlet-decorator-css-class>		</portlet-decorator>		<portlet-decorator id=\\"barebone\\" name=\\"Barebone\\">			<portlet-decorator-css-class>this-should-be-here</portlet-decorator-css-class>		</portlet-decorator>		<portlet-decorator id=\\"borderless\\" name=\\"Borderless\\">			<portlet-decorator-css-class>portlet-borderless</portlet-decorator-css-class>		</portlet-decorator>		<portlet-decorator id=\\"decorate\\" name=\\"Decorate\\">			<default-portlet-decorator>true</default-portlet-decorator>			<portlet-decorator-css-class>portlet-decorate</portlet-decorator-css-class>		</portlet-decorator>	</theme></look-and-feel>"`;

exports[`mergeJSON should merge look-and-feel json and output a valid look-and-feel json object 1`] = `
Object {
  "look-and-feel": Object {
    "compatibility": Array [
      Object {
        "version": Array [
          "7.0.0+",
        ],
      },
    ],
    "theme": Array [
      Object {
        "$": Object {
          "id": "base-theme",
          "name": "Base Theme",
        },
        "portlet-decorator": Array [
          Object {
            "$": Object {
              "id": "portlet-decorator-1",
              "name": "Decorator 1",
            },
            "portlet-decorator-css-class": Array [
              "portlet-decorator-1",
            ],
          },
          Object {
            "$": Object {
              "id": "portlet-decorator-2",
              "name": "Decorator 2",
            },
            "portlet-decorator-css-class": Array [
              "portlet-decorator-2",
            ],
          },
          Object {
            "$": Object {
              "id": "portlet-decorator-3",
              "name": "Decorator 3",
            },
            "portlet-decorator-css-class": Array [
              "portlet-decorator-3",
            ],
          },
        ],
        "settings": Array [
          Object {
            "setting": Array [
              Object {
                "$": Object {
                  "configurable": "true",
                  "key": "show-breadcrumb",
                  "type": "checkbox",
                  "value": "true",
                },
              },
              Object {
                "$": Object {
                  "configurable": "true",
                  "key": "parent-setting",
                  "type": "input",
                  "value": "true",
                },
              },
            ],
          },
        ],
        "template-extension": Array [
          "vm",
        ],
      },
    ],
  },
}
`;

exports[`mergeLookAndFeelJSON should merge look and feel json 1`] = `
Object {
  "look-and-feel": Object {
    "compatibility": Array [
      Object {
        "version": Array [
          "7.0.0+",
        ],
      },
    ],
    "theme": Array [
      Object {
        "$": Object {
          "id": "base-theme",
          "name": "Base Theme",
        },
        "portlet-decorator": Array [
          Object {
            "$": Object {
              "id": "portlet-decorator-1",
              "name": "Decorator 1",
            },
            "portlet-decorator-css-class": Array [
              "portlet-decorator-1",
            ],
          },
          Object {
            "$": Object {
              "id": "portlet-decorator-2",
              "name": "Decorator 2",
            },
            "portlet-decorator-css-class": Array [
              "portlet-decorator-2",
            ],
          },
          Object {
            "$": Object {
              "id": "portlet-decorator-3",
              "name": "Decorator 3",
            },
            "portlet-decorator-css-class": Array [
              "portlet-decorator-3",
            ],
          },
        ],
        "settings": Array [
          Object {
            "setting": Array [
              Object {
                "$": Object {
                  "configurable": "true",
                  "key": "show-breadcrumb",
                  "type": "checkbox",
                  "value": "true",
                },
              },
              Object {
                "$": Object {
                  "configurable": "true",
                  "key": "parent-setting",
                  "type": "input",
                  "value": "true",
                },
              },
            ],
          },
        ],
        "template-extension": Array [
          "vm",
        ],
      },
    ],
  },
}
`;
